### PR TITLE
DROOLS-3892: [DMN Designer] Data Types tab - Add a nested row to a new Structured data type by default

### PR DESCRIPTION
**Pull request by karreiro**
See: https://issues.jboss.org/browse/DROOLS-3892

---

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java [hunk 1]
```diff
@@ -423,6 +423,50 @@ public void testDoSaveAndCloseEditMode() {
         verify(dataTypeList).refreshItemsByUpdatedDataTypes(updatedDataTypes);
         verify(listItem).closeEditMode();
         verify(dataTypeList).fireOnDataTypeListItemUpdateCallback(newDataTypeHash);
+        verify(listItem).insertNewFieldIfDataTypeIsStructure();
+    }
+
+    @Test
+    public void testInsertNewFieldIfDataTypeIsStructureWhenDataTypeIsStructure() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: Could we please update method name. I think the name should contain just **IfDataTypeIsStructure** or **WhenDataTypeIsStructure**.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java [hunk 2]
```diff
@@ -423,6 +423,50 @@ public void testDoSaveAndCloseEditMode() {
...
+        doReturn(dataType).when(listItem).getDataType();
+        doNothing().when(listItem).insertNestedField();
+
+        listItem.insertNewFieldIfDataTypeIsStructure();
+
+        verify(listItem).insertNestedField();
+    }
+
+    @Test
+    public void testInsertNewFieldIfDataTypeIsStructureWhenDataTypeIsStructureButHasFields() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: Could we please update method name. I think the name should contain just **IfDataTypeIsStructure** or **WhenDataTypeIsStructure**.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java [hunk 3]
```diff
@@ -423,6 +423,50 @@ public void testDoSaveAndCloseEditMode() {
...
+        when(dataType.getSubDataTypes()).thenReturn(asList(mock(DataType.class), mock(DataType.class)));
+        doReturn(dataType).when(listItem).getDataType();
+
+        listItem.insertNewFieldIfDataTypeIsStructure();
+
+        verify(listItem, never()).insertNestedField();
+    }
+
+    @Test
+    public void testInsertNewFieldIfDataTypeIsStructureWhenDataTypeIsNotStructure() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: Could we please update method name. I think the name should contain just **IfDataTypeIsStructure** or **WhenDataTypeIsStructure**.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java (hunk 1 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: @jomarko I follow this pattern for method names when I'm testing: **test[METHOD_NAME]When[CONTEXT]**.

So, in this context, I think that makes sense the name **test[InsertNewFieldIfDataTypeIsStructure]When[DataTypeIsStructure]**.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java (hunk 2 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: @jomarko I follow this pattern for method names when I'm testing: **test[METHOD_NAME]When[CONTEXT]**.

So, in this context, I think that makes sense the name **test[InsertNewFieldIfDataTypeIsStructure]When[DataTypeIsStructureButHasFields]**.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java (hunk 3 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: @jomarko I follow this pattern for method names when I'm testing: **test[METHOD_NAME]When[CONTEXT]**.

So, in this context, I think that makes sense the name **test[InsertNewFieldIfDataTypeIsStructure]When[DataTypeIsNotStructure]**.

**Comment by karreiro** [comment]
Thank you for the review, @jomarko. Changes applied.

**Comment by manstis** [comment]
Jenkins please test this.

**Pull request merged**